File: services/printifyService.ts
import { ApiCredentials, PublishResult, LogType, LogEntry } from '../types';

const generateId = () => Math.random().toString(36).substr(2, 9);
const timestamp = () => new Date().toLocaleTimeString('en-US', { hour12: false });

type LogCallback = (entry: LogEntry) => void;

export class PrintifyService {
  private apiKey: string;
  private shopId: string;
  private baseUrl = 'https://api.printify.com/v1';

  constructor(credentials: ApiCredentials) {
    this.apiKey = credentials.printifyApiKey;
    this.shopId = credentials.printifyShopId;
  }

  /**
   * Create a product on Printify
   */
  async createProduct(
    designUrl: string,
    productTitle: string,
    blueprintId: number,
    onLog: LogCallback
  ): Promise<PublishResult> {
    try {
      onLog({
        id: generateId(),
        timestamp: timestamp(),
        message: `[Printify] Creating product: ${productTitle}`,
        type: LogType.INFO
      });

      // Simulate API delay
      await this.sleep(1500);

      // In production, this would make actual API call:
      // const response = await fetch(`${this.baseUrl}/shops/${this.shopId}/products.json`, {
      //   method: 'POST',
      //   headers: {
      //     'Authorization': `Bearer ${this.apiKey}`,
      //     'Content-Type': 'application/json'
      //   },
      //   body: JSON.stringify({
      //     title: productTitle,
      //     description: 'Auto-generated product from StaticWaves POD Studio',
      //     blueprint_id: blueprintId,
      //     print_provider_id: 1,
      //     variants: [{ /* variant config */ }],
      //     print_areas: [{ /* print area with design URL */ }]
      //   })
      // });

      const mockProductId = `pfy_${generateId()}`;

      onLog({
        id: generateId(),
        timestamp: timestamp(),
        message: `[Printify] ✓ Product created successfully (ID: ${mockProductId})`,
        type: LogType.SUCCESS
      });

      return {
        platform: 'printify',
        productId: mockProductId,
        productUrl: `https://printify.com/app/products/${mockProductId}`,
        status: 'success',
        message: 'Product created on Printify'
      };
    } catch (error) {
      onLog({
        id: generateId(),
        timestamp: timestamp(),
        message: `[Printify] ✗ Failed to create product: ${error}`,
        type: LogType.ERROR
      });

      return {
        platform: 'printify',
        status: 'failed',
        message: `Printify error: ${error}`
      };
    }
  }

  /**
   * Publish product to connected store (Shopify)
   */
  async publishToStore(productId: string, onLog: LogCallback): Promise<boolean> {
    try {
      onLog({
        id: generateId(),
        timestamp: timestamp(),
        message: `[Printify] Publishing product ${productId} to store...`,
        type: LogType.INFO
      });

      await this.sleep(1000);

      // In production:
      // await fetch(`${this.baseUrl}/shops/${this.shopId}/products/${productId}/publish.json`, {
      //   method: 'POST',
      //   headers: { 'Authorization': `Bearer ${this.apiKey}` }
      // });

      onLog({
        id: generateId(),
        timestamp: timestamp(),
        message: `[Printify] ✓ Product published to connected store`,
        type: LogType.SUCCESS
      });

      return true;
    } catch (error) {
      onLog({
        id: generateId(),
        timestamp: timestamp(),
        message: `[Printify] ✗ Failed to publish: ${error}`,
        type: LogType.ERROR
      });
      return false;
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

File: backend/services/printify_service.py
"""
Printify Integration Service
"""
import requests
from typing import List, Dict, Any, Optional
from loguru import logger

from config.settings import settings


class PrintifyService:
    """Service for Printify POD operations"""

    def __init__(self):
        self.base_url = "https://api.printify.com/v1"
        self.api_token = settings.PRINTIFY_API_TOKEN
        self.headers = {
            "Authorization": f"Bearer {self.api_token}",
            "Content-Type": "application/json",
        }

    async def create_product(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create POD product in Printify"""
        try:
            url = f"{self.base_url}/shops/{{shop_id}}/products.json"

            payload = {
                "title": data.get("title"),
                "description": data.get("description"),
                "blueprint_id": data.get("blueprint_id"),
                "print_provider_id": data.get("print_provider_id"),
                "variants": data.get("variants", []),
                "print_areas": data.get("print_areas", []),
            }

            response = requests.post(url, json=payload, headers=self.headers, timeout=30)
            response.raise_for_status()

            product = response.json()
            logger.info(f"Created Printify product: {product.get('id')}")

            return product

        except Exception as e:
            logger.error(f"Error creating Printify product: {e}")
            raise

    async def list_products(self) -> List[Dict[str, Any]]:
        """List Printify products"""
        try:
            url = f"{self.base_url}/shops/{{shop_id}}/products.json"

            response = requests.get(url, headers=self.headers, timeout=30)
            response.raise_for_status()

            data = response.json()
            products = data.get("data", [])

            return products

        except Exception as e:
            logger.error(f"Error listing Printify products: {e}")
            return []

    async def create_order(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create order in Printify"""
        try:
            url = f"{self.base_url}/shops/{{shop_id}}/orders.json"

            payload = {
                "external_id": data.get("external_id"),
                "line_items": data.get("line_items", []),
                "shipping_method": data.get("shipping_method", 1),
                "address_to": data.get("address_to"),
                "send_shipping_notification": True,
            }

            response = requests.post(url, json=payload, headers=self.headers, timeout=30)
            response.raise_for_status()

            order = response.json()
            logger.info(f"Created Printify order: {order.get('id')}")

            return order

        except Exception as e:
            logger.error(f"Error creating Printify order: {e}")
            raise

    async def get_blueprints(self) -> List[Dict[str, Any]]:
        """Get available product blueprints"""
        try:
            url = f"{self.base_url}/catalog/blueprints.json"

            response = requests.get(url, headers=self.headers, timeout=30)
            response.raise_for_status()

            blueprints = response.json()
            return blueprints

        except Exception as e:
            logger.error(f"Error getting blueprints: {e}")
            return []
